### PR TITLE
Update Development Environment - Support Node v14

### PR DESCRIPTION
**Pull request by astridx**
If I'm right, we don't fully support Node v12+ yet because of the way we integrate ECMAScript modules.

There are [two options](https://2ality.com/2019/04/nodejs-esm-impl.html#using-es-modules-on-node.js) since Node v12+: : Via library or via flag. As far as I can see, we use the library [esm](https://github.com/standard-things/esm).

By the way: Mapbox changed the way to use ECMAScript modules (esm) in PR 10367. They do not use  [esm](https://github.com/standard-things/esm) any more. 

~~I have no experience with esm. The last version is almost 2 years old. Nevertheless, I suggest updating to this one, because this way our Node scipts are usable under Node v12+. Not everything runs smoothly with this PR. For example, many tests fail. But maybe so we can better work on supporting Node v12+.~~

I decided to use ["type": "module"](https://nodejs.org/api/packages.html#packages_type) in main package.json because it is more [future-proof](https://nodejs.org/api/packages.html#packages_determining_module_system). 
For this reason, I have had to change `Node.js require` to` ES6 import/export` in many places. In addition, the specification of [file extensions](https://nodejs.org/api/esm.html#esm_mandatory_file_extensions) is mandatory. 

## Launch Checklist

 - [x] confirm your changes do not include backports from Mapbox projects (unless with compliant license)
 - [x] briefly describe the changes in this PR
 - [ ] write tests for all new functionality -> tests have to be correted
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

### CODE REVIEW

**Comment by HarelM** [comment]
Thanks for kicking this off!!
I guess we'll need to fix those tests before we can merge this in.
My two cents: if we can avoid using a library that seems no longer maintained I would feel better about this PR, but this can also be an initial step forward to mitigate.
There was also a suggestion to convert the code to typescript which will also solve this, wouldn't it? (Ton of work nevertheless).

**Comment by astridx** [comment]
> My two cents: if we can avoid using a library that seems no longer maintained I would feel better about this PR,

I see it exactly the same way. I thought about whether I should create the PR at all. 

I did it because I saw in Mapbox's PR that a lot has been changed to support Node v12+. If someone else can do this in one step, then that's great. Updating to the esm library creates the possibility to proceed in small steps.

I do not add the library. We are already using it. I'm just updating it. So far I haven't found any disadvantage. 
We would certainly have to adapt [that](https://github.com/maplibre/maplibre-gl-js/actions/runs/984047873) in any case, right?

**Comment by HarelM** [comment]
I've read through the issue related to the mapbox PR and they write there that this didn't work for them. Since I don't think muting failing tests is the right approach, and due to what they wrote it seems that this won't get us to where we want to be, I still think this pr is important to kick off the process, but I don't think we'll be able to easily fix this that way. I might be wrong. Feel free to commit more changes here until this is resolved...? Your call.
@nyurik, @wipfli any thoughts about this? Use this as a strong argument to migrate to TS? try solving this any other way?

**Comment by wipfli** [comment]
Thanks for working on this Astrid. Indeed, we cannot stay with node version 10 forever. Already now, gekodriver `v2` does not support node 10 so I had to restrict the driver to version 1. The sooner we go to a newer node version the better.

**Comment by wipfli** [comment]
Regarding the typescript discussion, I think there was already a long thread about this at some point. @HarelM do you remember where?

**Comment by HarelM** [comment]
Yes, the discussion is here: #32.
This node issues were never a point in the above discussion but I see correlation.

File: src/render/painter.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ import rasterBoundsAttributes from '../data/raster_bounds_attributes';
 import posAttributes from '../data/pos_attributes';
 import ProgramConfiguration from '../data/program_configuration';
 import CrossTileSymbolIndex from '../symbol/cross_tile_symbol_index';
-import * as shaders from '../shaders';
+import shaders from '../shaders/shaders.js';
```
Comment (HarelM) [inline src/render/painter.js:17]: Do we need the `.js` extension at the end? other files are not using this extension...

File: src/render/program.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // @flow
 
-import {prelude} from '../shaders';
+import {prelude} from '../shaders/shaders.js';
```
Comment (HarelM) [inline src/render/program.js:3]: Same as above...

Comment (astridx) [reply src/render/program.js]: Fixed. Thank you @HarelM

File: src/render/painter.js (hunk 1 above)
Comment (astridx) [reply src/render/painter.js]: Fixed. Thank you @HarelM

**Comment by HarelM** [comment]
Not sure how the ci works, but you'll need to update the circle ci file with node 14 instead of 10, currently seems like it continues to run everything on 10 so the ci is green...

**Comment by wipfli** [comment]
In my understanding CI is all done with GitHub actions.

**Comment by astridx** [comment]
@HarelM @wipfli 
I have changed the `circleci-yml` in one of the last commits. Should I not have done that?

So far I only look at the test results locally.

**Comment by HarelM** [comment]
You should change the circle-ci. It should be merged as part of this commit. Once this commit lands, the ci should be updated.

**Comment by HarelM** [comment]
Seems like an issue with yarn.lock...?

**Comment by astridx** [comment]
Now everything is red :(. Is that really correct?

Locally, I just started the unit tests. None of them failed. Not even the faulty `ScrollZoomHandler `unit test.

tests

I don't know if I can reach my goal with this approach. But the fact that everything is red makes me feel insecure more.

**Comment by astridx** [comment]
> Seems like an issue with yarn.lock...?

Hm. I think I add `yarn.lock`. This file is not in` .gitignore`. `package-lock.json` is in `,gitignore`.

**Comment by astridx** [comment]
> Seems like an issue with yarn.lock...?

Thank you. After deleting the changes in` yarn.lock` the ci-tests started again.

**Comment by HarelM** [comment]
I'm looking at the build which says:
Run yarn install --frozen-lockfile
yarn install v1.22.10
[1/5] Validating package.json...
error maplibre-gl@1.14.1-rc.2: The engine "node" is incompatible with this module. Expected version ">=14.16.0". Got "10.24.1"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

You'll get there, don't worry, it'll just take more time...

**Comment by HarelM** [comment]
I'm not a big fan of yarn, I guess I don't understand its added value over npm...

**Comment by HarelM** [comment]
Seems like two steps of the CI are failing - one is test-browser and the other timesout I guess... (test-query), I have no clue what's causing this to fail though, are these tests no failing locally?

**Comment by astridx** [comment]
As it is now, it matches my local installation. `test-browser` and `test-query` are not OK.

When all tests failed, it was because I had not run `yarn install`. I only called` npm ci`. Nevertheless, a `yarn.lock` was created. When I committed this` yarn.lock` to the github repo, all tests failed on github. Locally, only the two that are still faulty failed. 

Let's see if I can get these green too.

**Comment by HarelM** [comment]
Great! Thanks for tackling this!! If the CI env's results matches your env's results it's a good start :-)

**Comment by astridx** [comment]
> @astridx let me know if this would cover this one as well: #117
> Seems to me that both basically have the same purpose - so I prefer to keep the effort in one place.
> Keep me posted.

@HarelM I have seen that #117 is now closed. I have entered the changes from there in the same way here.

**Comment by astridx** [comment]
Now all tests are green here on Github. But on my local computer, at least `npm test-render` and `npm run test-query-node` still fail. Shouldn't these also be tested here on Github?

**Comment by HarelM** [comment]
I closed it in favor of this one... Good work!

**Comment by HarelM** [comment]
I don't know why they are not running...
In any case, you need to "undraft" this PR in order for it to be merged.
If this is green I think we are good to go. If we decide to add the tests I'm the future we can always make sure they are green before we do :-)

**Comment by HarelM** [comment]
Can you undo the change of the addition of ".js" extension to all the files? There must be a way without it...?

**Comment by astridx** [comment]
> I don't know why they are not running...
> In any case, you need to "undraft" this PR in order for it to be merged.
> If this is green I think we are good to go. If we decide to add the tests I'm the future we can always make sure they are green before we do :-)

I'll look at the two failing scripts tomorrow. If I find a solution to anything I notice, I'll change the draft status.



> Can you undo the change of the addition of ".js" extension to all the files? There must be a way without it...?

What disadvantage do you see with the added file extensions? Apart from the work that has now been done, I only see advantages. But maybe I'm missing something?

If I see it correctly, the only possibility to do without the file extensions is the use of https://github.com/standard-things/esm. I had started this solution, but run into problems during integration testing that I couldn't find a solution for. The Node documentation says that the extensions are mandatory: https://nodejs.org/api/esm.html#esm_mandatory_file_extensions

**Comment by HarelM** [comment]
Here are the problems the way I see it:
The number of file that needs to be reviewed is very large as this changes almost all the files, which will degrade the code review quality.
Typescript uses almost the same notation for imports and there's a discussion if this would be the next step for this library, so keeping it similar to how it will look in ts is a big plus, IMO.
less change will help future developers to better understand what changes really were in case we introduce a big in this PR.
If you'd ask me, I think it's important enough to try and avoid this change...

**Comment by HarelM** [comment]
Is it possible to add the extension only to non-flow-typed files? i.e. files that are part of the build and not part of the library?
Flow files get traspiled to js, so they are being taken care of, right?
I guess my question is if we can minimize the number of files that needs to be changed?

**Comment by astridx** [comment]
> Here are the problems the way I see it:
> The number of file that needs to be reviewed is very large as this changes almost all the files, which will degrade the code review quality.
> Typescript uses almost the same notation for imports and there's a discussion if this would be the next step for this library, so keeping it similar to how it will look in ts is a big plus, IMO.
> less change will help future developers to better understand what changes really were in case we introduce a big in this PR.

You are right, a review is big. 

About typescript: imports with file extensions work here too, right? 

I think that adding the extensions is more precise and thus also makes the code easier to read for future developers.

**Comment by astridx** [comment]
> Is it possible to add the extension only to non-flow-typed files? i.e. files that are part of the build and not part of the library?

If I see it correctly, it is only possible not to use file extensions when importing into node 12+ when using the `ems` package. 

I could keep trying to get this variant to work. I decided against it because the package was last changed in 2019 and node in its own doc describes the variant with `type:module`  in `package.json` as future proof. 
But maybe I am missing something.

**Comment by HarelM** [comment]
Don't get me wrong, using ems is not the right solution as it's kinda deprecated.
What I'm saying is that the js file that also have flow annotation are passing some traspilation which might allow keeping them as they are now without changing them.
Migration to typescript will change the file extension of all the files from .js to .ts and so I would like to keep the code un aware of file extensions.
I can fork your branch and play with it if you want...
I think it's possible according to things I read...

**Comment by astridx** [comment]
> I can fork your branch and play with it if you want...
> I think it's possible according to things I read...

This would be great. I can not figure it out but I like to learn.

**Comment by mourner** [comment]
@HarelM @astridx I'm extremely disappointed to find out this PR directly copied all the changes from my PR here https://github.com/mapbox/mapbox-gl-js/pull/10367 (not open-source-licensed) down to small details like the custom loader code. I worked extremely hard to get it working and sneakily and intentionally stealing my work directly is not fair. I urge you to close this PR and start from scratch on your own, without copy-pasting anything from GL JS, otherwise we would have explore what practices like this mean legally for this project.

**Comment by mourner** [comment]
image

@marcelnormann @nyurik please discuss this on the steering committee meeting. Note how the checkbox above was checked.

**Comment by astridx** [comment]
@mourner 
I'm sorry. I looked at these things primarily to learn. I had tried different things. I looked at your loader and it is currently in this PR. You are right.
I'll close here. It is difficult to change something that has already been done elsewhere in a similar way.

**Comment by mourner** [comment]
@astridx I may have overreacted and apologize for the harsh tone. It's fine to learn and then do similar changes independently (no objections to this at all), but definitely not OK to directly copy proprietary code.

**Comment by HarelM** [comment]
Thanks for the input @mourner.
I have nothing but respect to your work and the mapbox license, I'm sorry if you feel like we are copying your hard work. This was never the intention.
The checkbox above is exactly for that.
I'll take it to the committee to see what else can be done, but I'm not entirely sure how to address this as I have not looked into your code, I simply reviewed the changes here and with the objective of upgrading to latest version of node and other libraries, I believe ultimately the changes will be very similar - but I believe (haven't compared it obviously) that the commits and commit order and content is different.
In any case, we will start over per your request.

**Comment by mourner** [comment]
@HarelM for the record, leaving a few comparisons to demonstrate that some parts were copied as is, not just similar:

image
image

**Comment by mourner** [comment]
@astridx I now realize this was likely unintentional, so once again please forgive me for the initial angry reaction, and feel free to ask questions if you get stuck when redoing this, and I'll be happy to help. It should also be a lot much easier and faster to do it from scratch cleanly now after everything you've learned. Also, feel free to consult the list of gotchas I left in my PR description (but not the code changes). And you don't have to do the `.js` renames everywhere — Node seems to have an option to use the old path resolution logic that I wasn't aware of at the time.

**Comment by HarelM** [comment]
Thanks @mourner for your kind words and the offer for an olive branch.
I know that @astridx have worked very hard on this by the commits' times, scale, failing attempts and his comments in this conversation.
I think that if this was simply a copy-paste of your final code changes it would've been a lot simpler :-)
Having said that, I understand what you are saying about similar code changes. I'll need to think how I can prevent this in the future.
Sorry once again.

**Comment by ambientlight** [comment]
this calls for a github action or a bot that can cross checks PRs-vs-PRs for changes similarity via some plagiarism checker, not sure how many false positives can arise, but at least something that can post a warning comment to an open PR linking to PR on mapbox side would be handy

surprised there is nothing like that out there

**Comment by HarelM** [comment]
@ambientlight I generally agree this is the right approach. It'd be extremely cool if you could push this forward.

**Comment by marcelnormann** [comment]
> this calls for a github action or a bot that can cross checks PRs-vs-PRs for changes similarity via some plagiarism checker, not sure how many false positives can arise, but at least something that can post a warning comment to an open PR linking to PR on mapbox side would be handy

Please see #129. I already spent some time for research on that issue, but right now there is no real progress. Please let me know if you have some hints for us.

**Comment by marcelnormann** [comment]
> @astridx I may have overreacted and apologize for the harsh tone. It's fine to learn and then do similar changes independently (no objections to this at all), but definitely not OK to directly copy proprietary code.

As a steering committe it is up to us to apologize! Backporting of non-compliant code is unacceptable and against the rules of this project. We will take additional measures to ensure other peoples copyright.

**Pull request closed without merging**